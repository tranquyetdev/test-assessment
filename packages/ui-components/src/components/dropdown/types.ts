export interface DropdownProps {
  button: React.ReactNode;
  children: React.ReactNode;
}
