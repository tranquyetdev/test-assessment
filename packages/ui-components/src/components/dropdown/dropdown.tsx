import * as React from 'react';
import { DropdownProps } from './types';

export const Dropdown = React.forwardRef((props: DropdownProps, ref) => {
  const { button, children } = props;
  const [open, setOpen] = React.useState(false);
  const dropdown = React.useRef<HTMLInputElement>(null);

  React.useEffect(() => {
    const closeOpenMenus = (e: MouseEvent) => {
      if (
        dropdown &&
        dropdown.current &&
        !dropdown.current.contains(e.target as Node)
      ) {
        setOpen(false);
      }
    };

    document.addEventListener('mousedown', (e) => {
      closeOpenMenus(e);
    });
  }, []);

  const handleOpen = () => {
    setOpen(!open);
  };

  return (
    <div className="dropdown" ref={dropdown}>
      <button onClick={handleOpen} className="mb-[10px]">
        {button}
      </button>
      {open ? (
        <div className="bg-neutral-white shadow-md">{children}</div>
      ) : null}
    </div>
  );
});
