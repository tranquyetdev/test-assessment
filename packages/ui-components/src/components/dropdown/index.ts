export * from './dropdown';
export * from './types';
