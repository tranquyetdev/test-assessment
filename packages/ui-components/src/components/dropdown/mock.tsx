export const mockDropdownProps = {
  content: (
    <ul className="menu">
      <li className="menu-item">
        <button>Menu 1</button>
      </li>
      <li className="menu-item">
        <button>Menu 2</button>
      </li>
    </ul>
  ),
  button: <div className="border">Show content</div>,
};
