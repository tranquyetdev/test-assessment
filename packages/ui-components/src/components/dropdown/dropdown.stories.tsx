import type { Meta } from '@storybook/react';
import { Dropdown as DropdownComponent } from './dropdown';
import { mockDropdownProps } from './mock';

const Story: Meta<typeof DropdownComponent> = {
  title: 'Components / Dropdown',
  component: DropdownComponent,
};
export default Story;

export const Dropdown = () => (
  <DropdownComponent button={mockDropdownProps.button}>
    {mockDropdownProps.children}
  </DropdownComponent>
);
