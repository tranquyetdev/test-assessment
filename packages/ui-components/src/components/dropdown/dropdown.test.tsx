import React from 'react';
import { render } from '@testing-library/react';
import { Dropdown } from './dropdown';

describe('Dropdown', () => {
  Object.entries({}).forEach(([type, props]) => {
    it(`should render ${type} successfully`, () => {
      const { baseElement } = render(
        <Dropdown button={undefined} children={undefined} />
      );
      expect(baseElement).toMatchSnapshot();
    });
  });
});
