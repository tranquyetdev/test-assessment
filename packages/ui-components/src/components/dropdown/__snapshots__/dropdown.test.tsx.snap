// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render Dropdown successfully 1`] = `
<body>
  <div>
    <div
      class="dropdown"
    >
      <button
        class="mb-[10px]"
      >
        <div
          class="border"
        >
          Show content
        </div>
      </button>
    </div>
  </div>
</body>
`;
